### PR TITLE
Behaviour to install all airflow providers added

### LINKED ISSUE #11489: Make sure that local installation works with installing airflow + all providers.
We have to double-check that providers work and can be tested locally when installing airflow in a local virtualenv.

After we've changed to separate providers and core package, pip install . will not install those packages, but with local -e it will likely work as providers packages are in the same folder. 

We still need to check how local virtualenv work when we install without -e though.

Most likely solution will be dynamic selection of packages in setup.py based on the fact whether you are installing airflow from packages or sources. 

This is connected with #11464

### LINKED ISSUE #11490: Make sure Docker works with installing core + providers.
Currenlty when Docker is built, we are installing airflow using "pip install" from local sources. Since we split to separate providers, this process needs to be updated to also install relevant providers when certain extras are installed via PyPI and that it all work locally (there are the the separate, linked issues about it #11489 #11464 that should be implemented to cover this.

### PR DESCRIPTION
**Pull request by potiuk**
In Airflow 2.0 we decided to split Airlow into separate providers.
this means that when you prepare core airflow package, providers
are not installed by default. This is not very convenient for
local development though and for docker images built from sources,
where you would like to install all providers by default.

A new INSTALL_PROVIDERS_FROM_SOURCES environment variable controls
this behaviour now. It is is set to "true", all packages including
provider packages are installed. If missing or set to false, only
the core provider package is installed.

For Breeze, the default is set to "true", as for those cases you
want to install all providers in your environment. Similarly if you
build the production image from sources. However when you build
image using github tag or pip package, you should specify
appropriate extras to install the required provider packages.

Note that if you install Airflow via 'pip install .' from sources
in local virtualenv, provider packages are not going to be
installed unless you set INSTALL_PROVIDERS_FROM_SOURCES to "true".

Fixes #11489

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
Hey @ashb @kaxil  @turbaszek, others. I looked how to solve the installation problem described in #11489 and failing Kubernetes builds (caused likely by the provider split) and I believe i found the best approach. I think we want to keep this:

* in any setup where we use packages, we want to install separately airflow and providers
* in any setup where we use sources, we want to install airflow + providers together for development convenience

The proposal I have is a variable INSTALL_PROVIDERS_FROM_SOURCES. When this flag is "true", airflow will install providers from sources, when it is missing or anything else but "true", it will install only airflow core.  By default in Breeze and when building images from sources I set this variable to "true" (but it can be set to false by --skip-installing-airflow-providers flag so that you can also install a "bare" airflow in Breeze or prepare a "bare" image without any providers easily. Together with in-progess #11464 (I will have to add conditional dependencies there and rebase on top of this) it will have exactly the desired effect:

* when installing airflow with Breeze from sources (both prod and CI), we continue installing all providers from sources like we have in 1.10. Extras will not cause installing of apache-airflow-providers-* as those dependencies will be disabled in this case. You can also disable installing all providers by ``-skip-install0ing-airlfow-providers``

* when installing airflow with Breeze  from PyPI or GitHub, only the providers required by "extras" will be installed

* when installing airflow locally with "." (without -e) only bare Airflow will be installed, provders will only be available if you happen to be in the airflow directory or if you install provider packages manually. You can also set INSTALL_PROVIDER_SOURCES="true" before installation and then all providers will be installed as in 1.10.

* when installing airflow locally with -e '.'  - providers will be automatically installed and available. Since 'airflow' is taken direclty from sources, providers are there and they will be importable.


I tested all the above scenarios and I think it makes perfect sense. Let me know what you think.

File: IMAGES.rst [hunk 1]
```diff
@@ -390,6 +390,22 @@ The following build arguments (``--build-arg`` in docker build command) can be u
 |                                          |                                          | one of the folders included in           |
 |                                          |                                          | dockerignore                             |
```
Comment (potiuk) [inline IMAGES.rst]:
```suggestion
|                                          |                                          | dockeri gnore                            |
```

File: docs/production-deployment.rst [hunk 2]
```diff
@@ -387,6 +387,14 @@ The following build arguments (``--build-arg`` in docker build command) can be u
...
 |                                          |                                          | are pre-installed initially              |
 +------------------------------------------+------------------------------------------+------------------------------------------+
+| ``AIRFLOW_CONSTRAINTS_LOCATION``         |                                          | If not empty, it will override the       |
+|                                          |                                          | source of the constraints with the       |
+|                                          |                                          | specified URL or file. Note that the     |
+|                                          |                                          | file has to be in docker context so      |
+|                                          |                                          | it's best to place such file in          |
+|                                          |                                          | one of the folders included in           |
+|                                          |                                          | dockerignore                             |
```
Comment (potiuk) [inline docs/production-deployment.rst]:
```suggestion
|                                          |                                          | docker ignore                            |
```

**Review by mik-laj — APPROVED** [review]

File: IMAGES.rst [hunk 3]
```diff
@@ -388,7 +388,23 @@ The following build arguments (``--build-arg`` in docker build command) can be u
 |                                          |                                          | file has to be in docker context so      |
 |                                          |                                          | it's best to place such file in          |
 |                                          |                                          | one of the folders included in           |
-|                                          |                                          | dockerignore                             |
+|                                          |                                          | dockeri gnore                            |
```
Comment (kaxil) [inline IMAGES.rst]:
```suggestion
|                                          |                                          | dockerignore                            |
```

File: INSTALL [hunk 4]
```diff
@@ -42,6 +42,17 @@ python setup.py install
...
 
+By default `pip install` in Airflow 2.0 installs only the provider packages that are needed by the extras,
+however if you want to install all providers (which was default behaviour in 1.10.*)
+you can do it by setting environment variable INSTALL_PROVIDERS_FROM_SOURCES to `true`.
+
+INSTALL_PROVIDERS_FROM_SOURCES="true" pip install . \
+  --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-master/constraints-3.6.txt"
+
+
+You can also install airflow in development mode where not
```
Comment (kaxil) [inline INSTALL]: incomplete sentence

**Review by kaxil — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.rst [hunk 5]
```diff
@@ -520,9 +524,94 @@ yandexcloud, all, devel_ci
...
---------------------
+Airflow 2.0 is split into core and providers. They are delivered as separate packages:
+
+* apache-airflow: core
+* apache-airflow-providers-*: More than 50 provider packages
+
+In Airflow 1.10 all those providers were installed together within one single package and when you installed
+airflow locally, from sources, they were also installed. In Airflow 2.0, providers are separated out,
+and not installed together with the core, unless you set INSTALL_PROVIDERS_FROM_SOURCES environment
+variable to ``true``.
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
In Airflow 1.10 all those providers were installed together within one single package and when you installed
airflow locally, from sources, they were also installed. In Airflow 2.0, providers are separated out,
and not installed together with the core, unless you set ``INSTALL_PROVIDERS_FROM_SOURCES`` environment
variable to ``true``.
```

File: CONTRIBUTING.rst [hunk 6]
```diff
@@ -520,9 +524,94 @@ yandexcloud, all, devel_ci
...
+
+In Breeze - which is a development variable, INSTALL_PROVIDERS_FROM_SOURCES is set to true, but you
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
In Breeze - which is a development variable, ``INSTALL_PROVIDERS_FROM_SOURCES`` is set to true, but you
```

File: CONTRIBUTING.rst [hunk 7]
```diff
@@ -520,9 +524,94 @@ yandexcloud, all, devel_ci
...
+  .. END PACKAGE DEPENDENCIES HERE
+
+Backport providers
+------------------
+
+You can also build backport provider packages for Airflow 1.10. They aim to provide a bridge when users
+of Airflow 1.10 want to migrate to Airflow 2.0. The backport packages are named similarly to the
+provider packages, but with "backport" added:
+
+* 'apache-airflow-backport-provider-*
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
* 'apache-airflow-backport-provider-*'
```

**Review by kaxil — COMMENTED** [review]
Minor suggestions left

**Comment by potiuk** [comment]
All resolved @kaxil!

**Review by kaxil — APPROVED** [review]

**Pull request merged**